CHANGELOG.md

```diff
@@ -7,6 +7,8 @@
 ### Fixed
 * <How do the end-user experience this issue? what was the impact?> ([#????](https://github.com/realm/realm-core/issues/????), since v?.?.?)
 * Deleting an object in an asymmetric table would cause a crash. Likely to solve [#1537](https://github.com/realm/realm-kotlin/issues/1537), since v12.1.0.
+* Fixed FLX subscriptions not being fullfilled if the session was interrupted during bootstrapping. ([PR 7042](https://github.com/realm/realm-core/pull/7042))
+* Fixed FLX subscriptions not being picked up if an upload message was sent immediately after a subscription was committed but before the sync client checks for new subscriptions via `SubscriptionStore::get_next_pending_version()`. ([PR 7042](https://github.com/realm/realm-core/pull/7042))
```

Comment:
> I would be interested to see some logs when this happens. I noticed the snapshot value issue before, but as far as I could tell it wasn't causing any problems.

On the newer BAAS version, this was reproducible in the `app: flx-sync basic tests` test. Here are the truncated logs before the fix, showing the series of events leading to a hang waiting on the fulfillment of a subscription:

<details>

```
DB: 4193 Thread 0x16d31b000: subscription version 3 updating state from AwaitingMark to Complete
DB: 4193 Thread 0x16d31b000: Query find all: 'version < 3', limit = -1
DB: 4193 Thread 0x16d31b000: Query found: 1, Duration: 12 us
DB: 4193 Thread 0x16d31b000: Initiate commit version: 26
DB: 4193 Thread 0x16d31b000: db state for version 26 is: 
{
"sync_internal_schemas":[{"_key":0,"schema_group_name":"flx_subscription_store","schema_version":2},{"_key":1,"schema_group_name":"pending_bootstraps","schema_version":1}]
,"flx_subscription_sets":[{"_key":3,"version":3,"state":3,"snapshot_version":21,"error":null,"subscriptions":[{"_key":3,"id":"6531b6e2fcbfea88cbae7b8c","created_at":"2023-10-19 23:08:18.980461000","updated_at":"2023-10-19 23:08:19.049395000","name":null,"object_class":"Obj","query":"name == \"bar\" "}]}]
,"metadata":[{"_key":0,"version":1}]
,"class_Obj":[{"_key":1,"_id":"6531b6e0fcbfea88cbae7b89","name":"bar","value":10}]
,"flx_pending_bootstrap":[]
}


DB: 4193 Thread 0x16d31b000: Commit of size 3648 done in 1889 us ref 12608
DB: 4193 Thread 0x16d31b000: writemutex released
DB: 4193 Thread 0x202852080: End transaction 1869
DB: 4193 Thread 0x16d31b000: Start frozen 1796: 26 ref 12608
DB: 4193 Thread 0x202852080: Tr 1709: Advance read: 21 -> 26 ref 12608
DB: 4193 Thread 0x16d31b000: End transaction 1796
DB: 4193 Thread 0x16d31b000: End transaction 1981
DB: 4193 Thread 0x202852080: Start frozen 2034: 26 ref 12608
DB: 4193 Thread 0x16d31b000: Start read 1796: 26 ref 12608
DB: 4193 Thread 0x16d31b000: End transaction 1796
DB: 4193 Thread 0x16d31b000: Start read 1796: 26 ref 12608
DB: 4193 Thread 0x16d31b000: End transaction 1796
DB: 4193 Thread 0x202852080: End transaction 2034
DB: 4193 Thread 0x202852080: acquire writemutex
DB: 4193 Thread 0x202852080: writemutex acquired
DB: 4193 Thread 0x202852080: Start write 1981: 26 ref 12608
DB: 4193 Thread 0x202852080: Initiate commit version: 27
DB: 4193 Thread 0x202852080: db state for version 27 is: 
{
"sync_internal_schemas":[{"_key":0,"schema_group_name":"flx_subscription_store","schema_version":2},{"_key":1,"schema_group_name":"pending_bootstraps","schema_version":1}]
,"flx_subscription_sets":[{"_key":3,"version":3,"state":3,"snapshot_version":21,"error":null,"subscriptions":[{"_key":3,"id":"6531b6e2fcbfea88cbae7b8c","created_at":"2023-10-19 23:08:18.980461000","updated_at":"2023-10-19 23:08:19.049395000","name":null,"object_class":"Obj","query":"name == \"bar\" "}]},{"_key":4,"version":4,"state":1,"snapshot_version":26,"error":null,"subscriptions":[]}]
,"metadata":[{"_key":0,"version":1}]
,"class_Obj":[{"_key":1,"_id":"6531b6e0fcbfea88cbae7b89","name":"bar","value":10}]
,"flx_pending_bootstrap":[]
}


DB: 4193 Thread 0x16d31b000: Start read 1796: 26 ref 12608
DB: 4193 Thread 0x202852080: Commit of size 2424 done in 932 us ref 10432
DB: 4193 Thread 0x16d31b000: Query find all: 'version > 3 and (state == 1 or state == 2) and snapshot_version > 24', limit = -1
DB: 4193 Thread 0x202852080: writemutex released
DB: 4193 Thread 0x16d31b000: Query found: 0, Duration: 31 us
DB: 4193 Thread 0x16d31b000: End transaction 1796
DB: 4193 Thread 0x202852080: Start frozen 1796: 27 ref 10432
DB: 4193 Thread 0x16d31b000: Start read 2034: 27 ref 10432
DB: 4193 Thread 0x202852080: End transaction 1796
DB: 4193 Thread 0x16d31b000: End transaction 2034
Connection[2]: Session[2]: Sending: UPLOAD(progress_client_version=27, progress_server_version=24, locked_server_version=24, num_changesets=0)
DB: 4193 Thread 0x16d31b000: Start read 2034: 27 ref 10432
DB: 4193 Thread 0x16d31b000: End transaction 2034
DB: 4193 Thread 0x16d31b000: Start read 2034: 27 ref 10432
DB: 4193 Thread 0x16d31b000: End transaction 2034
DB: 4193 Thread 0x16d31b000: Start read 1796: 27 ref 10432

// NOTE: the following log is from Session::send_message(). The method get_next_pending_version does not find 
// the just committed subscription because the immediately previous upload just ran on version 27. 
//   m_pending_flx_sub_set = get_flx_subscription_store()->get_next_pending_version(
//   m_last_sent_flx_query_version, m_upload_progress.client_version);

DB: 4193 Thread 0x16d31b000: Query find all: 'version > 3 and (state == 1 or state == 2) and snapshot_version > 26', limit = -1
DB: 4193 Thread 0x16d31b000: Query found: 0, Duration: 34 us
DB: 4193 Thread 0x16d31b000: End transaction 1796
Connection[2]: Sending: PING(timestamp=2088136071, rtt=0)
Connection[2]: Received: PONG(timestamp=2088136071)
Connection[2]: Round trip time was 2 milliseconds
Connection[2]: Will emit a ping in 54487 milliseconds

// Hang forever waiting for subscription to be fulfilled. 
```

</details>

Reply:
IIUC the issue is that the UPLOAD with client version 27 is sent, but before that it should have sent the QUERY message (version 27 is the result of the subscription being commited, so upload is aware of the new subscriptions but somehow lost track of it). I guess the subscription is actually never sent to the server because the sync client gets past it. Can you try reverting the change at https://github.com/realm/realm-core/commit/ddafdbdc0bd179db627af82c1d8bf4ae0e3eac2a and run the test again?

Reply:
Interesting. The test does pass with that commit reverted.

Reply:
What do think is best, committing the correct writer version number, or reintroducing ddafdbd ?

Reply:
I think there may be another way to fix it 🤔